### PR TITLE
Fixed join multiple guilds exploit

### PR DESCRIPTION
**Pull request by IntelligentQuantum**
OLD PR: https://github.com/azerothcore/azerothcore-wotlk/pull/1789

Tested And It works fine for me

##### HOW TO TEST THE CHANGES:

First: Clone WowAddin (WoW Console) from https://github.com/tomrus88/WowAddin
Then: Use below code as a method and register the command in order to function

```
bool GuildLeave(char const*, char const* args)
{
  CDataStore Packet(CMSG_GUILD_DECLINE);
  Packet.Finalize();
  ClientServices::SendPacket(&Packet);

  Console::Write("Packet Sent", ECHO_COLOR);
  return true;
}
```

After That: Try to compile and Inject it to Wow.exe Process using a DLL Injector

Then see this video without this PR: https://vimeo.com/594412340

### CODE REVIEW

**Comment by lineagedr** [comment]
Why not just do https://github.com/azerothcore/azerothcore-wotlk/blob/master/src/server/game/Handlers/GuildHandler.cpp#L79

```diff
LOG_DEBUG("guild", "CMSG_GUILD_DECLINE [%s]", GetPlayerInfo().c_str());

+if (GetPlayer()->GetGuild())
+    return;

GetPlayer()->SetGuildIdInvited(0);
```

**Comment by IntelligentQuantum** [comment]
> Why not just do https://github.com/azerothcore/azerothcore-wotlk/blob/master/src/server/game/Handlers/GuildHandler.cpp#L79
> 
> ```diff
> LOG_DEBUG("guild", "CMSG_GUILD_DECLINE [%s]", GetPlayerInfo().c_str());
> 
> +if (GetPlayer()->GetGuild())
> +    return;
> 
> GetPlayer()->SetGuildIdInvited(0);
> ```

:+1: Yeah why not, i'm going to test this

**Comment by IntelligentQuantum** [comment]
> Why not just do https://github.com/azerothcore/azerothcore-wotlk/blob/master/src/server/game/Handlers/GuildHandler.cpp#L79
> 
> ```diff
> LOG_DEBUG("guild", "CMSG_GUILD_DECLINE [%s]", GetPlayerInfo().c_str());
> 
> +if (GetPlayer()->GetGuild())
> +    return;
> 
> GetPlayer()->SetGuildIdInvited(0);
> ```

it works fine, Can you make PR ?

File: src/server/game/Guilds/Guild.cpp [hunk 1]
```diff
@@ -2197,6 +2197,15 @@ bool Guild::AddMember(ObjectGuid guid, uint8 rankId)
...
         rankId = _GetLowestRankId();
 
+    ObjectGuid::LowType guidLow = guid.GetCounter();
+
+    CharacterDatabasePreparedStatement* memberGuidQuery = CharacterDatabase.GetPreparedStatement(CHAR_SEL_GUILD_MEMBER_GUID);
+    memberGuidQuery->setUInt32(0, guidLow);
+    PreparedQueryResult memberGuidResult = CharacterDatabase.Query(memberGuidQuery);
+
+    if (memberGuidResult)
+        return false;
```
Comment (FrancescoBorzi) [inline src/server/game/Guilds/Guild.cpp:2207]:
```suggestion
    if (memberGuidResult)
    {
        return false;
    }
```

**Comment by FrancescoBorzi** [comment]
@IntelligentQuantum thanks for this PR.

Please don't forget to:

- specify what is the status of it and apply the Label(s) accordingly (waiting to be tested? ready to be reviewed?)
- also specify whether you tested this in-game already or not

**Comment by lineagedr** [comment]
Correct me if I'm wrong but do we even need GetPlayer()->SetInGuild(0); ?

So if we decline a guild invite it will set our current guildId to 0 which doesn't make any sense we should either remove it
or check if we're not in a guild and then set it to 0 which again wouldn't make any sense because we're not in a guild removing it might be the best option
```c++
void WorldSession::HandleGuildDeclineOpcode(WorldPacket& /*recvPacket*/)
{
    LOG_DEBUG("guild", "CMSG_GUILD_DECLINE [%s]", GetPlayerInfo().c_str());
    GetPlayer()->SetGuildIdInvited(0);
}
```

**Comment by IntelligentQuantum** [comment]
@lineagedr I removed this in old PR https://github.com/azerothcore/azerothcore-wotlk/pull/1789/files

**Comment by lineagedr** [comment]
@IntelligentQuantum Any reason that you reverted that fix? Because I don't see any harm in removing it

**Comment by IntelligentQuantum** [comment]
@lineagedr They said it has crash, But i have used it from almost 5 years and there is no problem with it.

**Comment by FrancescoBorzi** [comment]
@IntelligentQuantum is this ready to be tested?

**Comment by IntelligentQuantum** [comment]
> @IntelligentQuantum is this ready to be tested?

Yes

File: src/server/game/Handlers/GuildHandler.cpp [hunk 2]
```diff
@@ -77,6 +77,9 @@ void WorldSession::HandleGuildDeclineOpcode(WorldPacket& /*recvPacket*/)
 {
     LOG_DEBUG("guild", "CMSG_GUILD_DECLINE [%s]", GetPlayerInfo().c_str());
 
+    if (GetPlayer()->GetGuild())
+        return;
```
Comment (Helias) [inline src/server/game/Handlers/GuildHandler.cpp:81]:
```suggestion
    if (GetPlayer()->GetGuild())
    {
        return;
    }
```

**Review by Helias — APPROVED** [review]

**Comment by FrancescoBorzi** [comment]
@IntelligentQuantum if you tested the normal guild actions and they still work fine (e.g. try to create a guild, invite someone, remove someone, etc...), I'd say we can merge this.

**Pull request merged**